src/internal/connector/onedrive/restore.go

```diff
@@ -541,43 +562,112 @@ func restoreV6File(
 	return itemInfo, nil
 }
 
+// createRestoreFoldersWithPermissions creates the restore folder hierarchy in
+// the specified drive and returns the folder ID of the last folder entry in the
+// hierarchy. Permissions are only applied to the last folder in the hierarchy.
+// Passing nil for the permissions results in just creating the folder(s).
+// folderCache is mutated, as a side effect of populating the items.
+func createRestoreFoldersWithPermissions(
+	ctx context.Context,
+	creds account.M365Config,
+	service graph.Servicer,
+	drivePath *path.DrivePath,
+	driveRootID string,
+	restoreFolders *path.Builder,
+	folderPath path.Path,
+	folderMetadata Metadata,
+	folderMetas map[string]Metadata,
+	fc *folderCache,
+	permissionIDMappings map[string]string,
+	restorePerms bool,
+) (string, error) {
+	id, err := CreateRestoreFolders(
+		ctx,
+		service,
+		drivePath.DriveID,
+		driveRootID,
+		restoreFolders,
+		fc)
+	if err != nil {
+		return "", err
+	}
+
+	if len(drivePath.Folders) == 0 {
+		// No permissions for root folder
+		return id, nil
+	}
+
+	if !restorePerms {
+		return id, nil
+	}
+
+	err = RestorePermissions(
+		ctx,
+		creds,
+		service,
+		drivePath.DriveID,
+		id,
+		folderPath,
+		folderMetadata,
+		folderMetas,
+		permissionIDMappings)
+
+	return id, err
+}
+
 // CreateRestoreFolders creates the restore folder hierarchy in the specified
 // drive and returns the folder ID of the last folder entry in the hierarchy.
+// folderCache is mutated, as a side effect of populating the items.
 func CreateRestoreFolders(
 	ctx context.Context,
 	service graph.Servicer,
-	driveID string,
-	restoreFolders []string,
+	driveID, driveRootID string,
```

Comment:
Is the driveRootID addition here required or is that an optimization? It looks like the latter but want to confirm I'm reading the PR right?

Reply:
Prefer we move that to a separate PR if indeed it's an optimization - there's a lot of lines of code changes for that.

Reply:
It's required in order to look up/create the top folder.

Reply:
Hmm - I'm missing something but why can't we continue to call `service.Client().DrivesById(driveID).Root().Get(ctx, nil)` in this method?

Reply:
We could, I moved it to avoid redundancy.  We're calling it N times here (where N is equal to folder count), when we only need to call it once.  Not so much a performance concern as it is recognizing the behavior of the world.